From 78d18c702ff2476d28301d55a51ba1eba2c0fc12 Mon Sep 17 00:00:00 2001
From: Himanshu Rawat <rwt@google.com>
Date: Wed, 15 Oct 2025 10:08:37 -0700
Subject: [PATCH 1/5] RFCOMM MUX connection request must not start bonding

Outgoing connection requests for secure sockets adds the bonding
security requirements for RFCOMM MUX. This security requirement is reset
only when an outgoing insecure socket connection is requested. This
means that all subsequent incoming connection requests for RFCOMM MUX
also enforce the bonding requirement. So incoming RFCOMM connection
leads to initiating bonding.

This is a backport of ag/36338174 required for a security fix. Also
includes flag removal for upgrade_temp_bonding_on_auth_req.

Test: atest BumbleBluetoothTests:RfcommTest
Flag: com.android.bluetooth.flags.upgrade_temp_bonding_on_auth_req
Bug: 385747513
Bug: 380091558
Bug: 447000848
Change-Id: Ia5f746e3036676faa49d2b19d11be48e5b7fea0f
(cherry picked from commit 3cce3201a64ebe5806fa0842ea0475ee68ed8910)
---
 flags/security.aconfig      | 10 ----------
 system/bta/jv/bta_jv_act.cc |  7 -------
 system/stack/btm/btm_sec.cc |  4 ++--
 3 files changed, 2 insertions(+), 19 deletions(-)

diff --git a/flags/security.aconfig b/flags/security.aconfig
index 8d5a8c1f31..1489199535 100644
--- a/flags/security.aconfig
+++ b/flags/security.aconfig
@@ -78,16 +78,6 @@ flag {
     }
 }
 
-flag {
-    name: "upgrade_temp_bonding_on_auth_req"
-    namespace: "bluetooth"
-    description: "Upgrade security If it is on temp bonding & authentication is requirement"
-    bug: "385747513"
-    metadata {
-        purpose: PURPOSE_BUGFIX
-    }
-}
-
 flag {
     name: "disconnect_reason_for_encryption_failure"
     namespace: "bluetooth"
diff --git a/system/bta/jv/bta_jv_act.cc b/system/bta/jv/bta_jv_act.cc
index d9eb9d0bbc..715c5a2311 100644
--- a/system/bta/jv/bta_jv_act.cc
+++ b/system/bta/jv/bta_jv_act.cc
@@ -1526,13 +1526,6 @@ void bta_jv_rfcomm_connect(tBTA_SEC sec_mask, uint8_t remote_scn, const RawAddre
                   },
   };
 
-  // Update security service record for RFCOMM client so that
-  // secure RFCOMM connection will be authenticated with MTIM protection
-  // while creating the L2CAP connection.
-  get_btm_client_interface().security.BTM_SetSecurityLevel(true, "RFC_MUX", BTM_SEC_SERVICE_RFC_MUX,
-                                                           sec_mask, BT_PSM_RFCOMM,
-                                                           BTM_SEC_PROTO_RFCOMM, 0);
-
   port_status = RFCOMM_CreateConnectionWithSecurity(UUID_SERVCLASS_SERIAL_PORT, remote_scn, false,
                                                     BTA_JV_DEF_RFC_MTU, peer_bd_addr, &handle,
                                                     bta_jv_port_mgmt_cl_cback, sec_mask, cfg);
diff --git a/system/stack/btm/btm_sec.cc b/system/stack/btm/btm_sec.cc
index 4595794cab..9b25e962a1 100644
--- a/system/stack/btm/btm_sec.cc
+++ b/system/stack/btm/btm_sec.cc
@@ -1366,8 +1366,8 @@ static bool btm_sec_is_upgrade_possible(tBTM_SEC_DEV_REC* p_dev_rec, bool is_ori
 
     /*if authentication is requirement & currently on temp bonding
      * trigger pairing */
-    if (com::android::bluetooth::flags::upgrade_temp_bonding_on_auth_req() &&
-        (p_dev_rec->sec_rec.security_required & BTM_SEC_OUT_AUTHENTICATE) &&
+    if ((p_dev_rec->sec_rec.security_required &
+         (is_originator ? BTM_SEC_OUT_AUTHENTICATE : BTM_SEC_IN_AUTHENTICATE)) &&
         p_dev_rec->sec_rec.is_bond_type_temporary()) {
       is_possible = true;
     }
-- 
2.53.0

